api/provider.go

```diff
@@ -0,0 +1,190 @@
+package api
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"strings"
+
+	"golang.org/x/oauth2"
+	"golang.org/x/oauth2/bitbucket"
+	"golang.org/x/oauth2/github"
+)
+
+// Provider is an interface for interacting with external account providers
+type Provider interface {
+	GetUserEmail(context.Context, *oauth2.Token) (string, error)
+	GetOAuthToken(context.Context, string) (*oauth2.Token, error)
+}
+
+var knownProviders = map[string]func(key, secret string) Provider{
+	"github":    NewGithubProvider,
+	"bitbucket": NewBitbucketProvider,
+}
+
+// Provider returns a Provider inerface for the given name
+func (a *API) Provider(name string) (Provider, error) {
+	fn, ok := knownProviders[name]
+	if !ok {
+		return nil, errors.New("Unknown provider")
+	}
+
+	name = strings.ToLower(name)
+
+	var key, secret string
+	switch name {
+	case "github":
+		key, secret = a.config.External.Github.Key, a.config.External.Github.Secret
+	case "bitbucket":
+		key, secret = a.config.External.Bitbucket.Key, a.config.External.Bitbucket.Secret
+	}
+
+	if key == "" || secret == "" {
+		return nil, fmt.Errorf("Provider %s could not be found", name)
+	}
+
+	return fn(key, secret), nil
+}
+
+// Github
+
+type githubProvider struct {
+	*oauth2.Config
+}
+
+func NewGithubProvider(key, secret string) Provider {
+	return &githubProvider{
+		&oauth2.Config{
+			ClientID:     key,
+			ClientSecret: secret,
+			Endpoint:     github.Endpoint,
+		},
+	}
+}
+
+func (g githubProvider) GetOAuthToken(ctx context.Context, code string) (*oauth2.Token, error) {
+	return g.Exchange(ctx, code)
+}
+
+func (g githubProvider) GetUserEmail(ctx context.Context, tok *oauth2.Token) (string, error) {
+	client := g.Client(ctx, tok)
+	res, err := client.Get("https://api.github.com/user/emails")
+	if err != nil {
+		return "", err
+
+	}
+	defer res.Body.Close()
+
+	userEmail := ""
+	emails := []map[string]interface{}{}
```

Comment:
instead of using a `map[string]interface{}` you could declare an anoymous struct and let the json lib do the work for you:

https://play.golang.org/p/12HbMfJT3V

In that case you could also try and extract all the common logic around requesting and parsing the result into a common method that you pass in the client, endpoint and resulting struct.